File: shared/src/commonMain/kotlin/com/kencric/scoring/ui/screens/ux04creatematch/CreateMatchForm.kt
package com.kencric.scoring.ui.screens.ux04creatematch

import com.kencric.scoring.core.ports.IdPort

/**
 * TASK-0041: `ux-specification.md UX-04`'s own logic, ported field-for-
 * field from `apps/web/src/screens/UX-04-create-match/createMatchForm.ts`
 * (`TASK-0040`) -- cross-platform parity (`C-7`) means this Kotlin
 * version and that TypeScript version must produce identical outputs
 * for identical inputs, verified here by mirroring `TASK-0040`'s own
 * test cases exactly, input for input, expected output for expected
 * output.
 *
 * HONEST SCOPE NOTE, confirmed with the user before writing this file:
 * no Android SDK, Gradle, or Kotlin toolchain exists anywhere in this
 * session (confirmed since `TASK-0016`, unchanged) -- there is no
 * `apps/android/` Gradle project to place this in, and nothing here
 * can be compiled, run, or verified beyond hand-tracing. Placed in
 * `shared/src/commonMain`, not a fabricated `apps/android/` path,
 * because this logic is genuinely platform-independent (no Android API
 * surface touched at all -- pure form state/validation) and this is
 * exactly what Kotlin Multiplatform exists for: one implementation a
 * real Android screen would eventually consume directly. Unlike the
 * web side (which needed its own separate TypeScript port because no
 * compiled Kotlin/JS bundle exists), a real `apps/android/` project
 * could import this file completely unchanged once one exists --
 * substitution, not a rewrite.
 */

enum class MatchFormat { T20, ODI_LIST_A, T10, THE_HUNDRED, CUSTOM }

sealed class Ownership {
    object Guest : Ownership()
    data class Organization(val organizationId: String) : Ownership()
}

data class MatchTemplate(val id: String, val organizationId: String, val label: String)

data class CreateMatchFormState(
    val matchLabel: String,
    val ownership: Ownership,
    val selectedTemplateId: String? = null,
    val format: MatchFormat? = null,
)

fun initialFormState(suggestedLabel: String, ownership: Ownership): CreateMatchFormState =
    CreateMatchFormState(matchLabel = suggestedLabel, ownership = ownership)

/** UX-04's own Validation rule: "A format must be chosen before Continue enables." */
fun canContinue(state: CreateMatchFormState): Boolean = state.format != null

/**
 * UX-04's own Validation rule: "if a selected template belongs to a
 * different organization than the current context, it's flagged
 * before use." Uses an exhaustive `when` over the sealed [Ownership]
 * rather than an early-return `is` check followed by relying on smart-
 * cast narrowing to the only other subtype -- deliberately, since
 * there is no compiler here to confirm that narrowing actually holds;
 * the explicit `when` is unambiguous either way.
 */
fun templateOrganizationMismatch(state: CreateMatchFormState, templates: List<MatchTemplate>): Boolean {
    val templateId = state.selectedTemplateId ?: return false
    return when (val ownership = state.ownership) {
        is Ownership.Guest -> false
        is Ownership.Organization -> {
            val template = templates.find { it.id == templateId }
            template != null && template.organizationId != ownership.organizationId
        }
    }
}

/**
 * `data-specification.md §5.1`'s `matches` table requires
 * `home_team_id`/`away_team_id`/`match_timezone` (all `NOT NULL`),
 * none of which UX-04 itself collects (`TASK-0040`'s own flagged
 * finding, carried over here identically) -- "Continue" produces a
 * local draft, not a call to `TASK-0039`'s `createMatch()`.
 */
data class DraftMatch(
    val id: String,
    val matchLabel: String,
    val organizationId: String?,
    val format: MatchFormat,
    val templateId: String?,
)

fun continueFromCreateMatch(state: CreateMatchFormState, idPort: IdPort): DraftMatch? {
    if (!canContinue(state)) return null
    val format = state.format ?: return null // canContinue already guarantees non-null; re-checked here only to satisfy the compiler's null-flow analysis, matching the TS version's `as MatchFormat` cast after its own equivalent guard.
    val organizationId = when (val ownership = state.ownership) {
        is Ownership.Guest -> null
        is Ownership.Organization -> ownership.organizationId
    }
    return DraftMatch(
        id = idPort.newId(),
        matchLabel = state.matchLabel,
        organizationId = organizationId,
        format = format,
        templateId = state.selectedTemplateId,
    )
}

sealed class TemplateListState {
    object Loading : TemplateListState()
    data class Populated(val templates: List<MatchTemplate>, val isFromCache: Boolean) : TemplateListState()
    object Empty : TemplateListState()
    data class FetchFailedFallbackToCache(val templates: List<MatchTemplate>) : TemplateListState()
    object FetchFailedNoCache : TemplateListState()
}

/** UX-04's own Error handling + Empty states, unified: "Template list fetch failure falls back to the last-synced local cache silently if any exist, otherwise offers 'Continue with custom setup' so the flow is never blocked." */
fun resolveTemplateListState(
    fetchSucceeded: Boolean,
    fetchedTemplates: List<MatchTemplate>?,
    cachedTemplates: List<MatchTemplate>,
): TemplateListState {
    if (fetchSucceeded && fetchedTemplates != null) {
        return if (fetchedTemplates.isEmpty()) {
            TemplateListState.Empty
        } else {
            TemplateListState.Populated(fetchedTemplates, isFromCache = false)
        }
    }
    if (cachedTemplates.isNotEmpty()) return TemplateListState.FetchFailedFallbackToCache(cachedTemplates)
    return TemplateListState.FetchFailedNoCache
}

File: apps/web/src/screens/UX-04-create-match/createMatchForm.ts
/**
 * TASK-0040: `ux-specification.md UX-04`'s own logic, kept as plain
 * TypeScript separate from the React component (`CreateMatchScreen.tsx`)
 * so it is directly, cheaply unit-testable without rendering anything --
 * the same "test the logic, not just the render" discipline this
 * session has applied throughout `shared/`.
 */

import type { IdPort } from "../../core/sharedCoreStub";

export type MatchFormat = "T20" | "ODI_LIST_A" | "T10" | "THE_HUNDRED" | "CUSTOM";

export type Ownership = "GUEST" | { organizationId: string };

export interface MatchTemplate {
  id: string;
  organizationId: string;
  label: string;
}

export interface CreateMatchFormState {
  matchLabel: string;
  ownership: Ownership;
  selectedTemplateId: string | null;
  format: MatchFormat | null;
}

export function initialFormState(suggestedLabel: string, ownership: Ownership): CreateMatchFormState {
  return { matchLabel: suggestedLabel, ownership, selectedTemplateId: null, format: null };
}

/** UX-04's own Validation rule: "A format must be chosen before Continue enables." */
export function canContinue(state: CreateMatchFormState): boolean {
  return state.format !== null;
}

/** UX-04's own Validation rule: "if a selected template belongs to a different organization than the current context, it's flagged before use." */
export function templateOrganizationMismatch(state: CreateMatchFormState, templates: readonly MatchTemplate[]): boolean {
  if (state.selectedTemplateId === null) return false;
  if (state.ownership === "GUEST") return false;
  const template = templates.find((t) => t.id === state.selectedTemplateId);
  if (!template) return false;
  return template.organizationId !== state.ownership.organizationId;
}

/**
 * `data-specification.md §5.1`'s `matches` table requires
 * `home_team_id`/`away_team_id`/`match_timezone` (all `NOT NULL`) --
 * none of which UX-04 itself collects (team selection is `UX-06`,
 * timezone is `UX-05`). "Continue" therefore CANNOT call `TASK-0039`'s
 * `createMatch()` yet -- doing so would mean fabricating values this
 * screen never had. It produces a `DraftMatch` instead, carrying
 * forward exactly what this screen collected, with a client-generated
 * id (matching `matches.id`'s own "Client-generated -- a match is
 * always created offline-first" contract), for whichever later
 * screen/task assembles the full create payload once team selection
 * and match setup are also complete. This is a genuine, flagged open
 * integration question between `UX-04`-`09`'s multi-screen flow and
 * the single-call generic-CRUD `createMatch()` `TASK-0039` built --
 * not silently resolved here by guessing.
 */
export interface DraftMatch {
  id: string;
  matchLabel: string;
  organizationId: string | null;
  format: MatchFormat;
  templateId: string | null;
}

export function continueFromCreateMatch(state: CreateMatchFormState, idPort: IdPort): DraftMatch | null {
  if (!canContinue(state)) return null;
  return {
    id: idPort.newId(),
    matchLabel: state.matchLabel,
    organizationId: state.ownership === "GUEST" ? null : state.ownership.organizationId,
    format: state.format as MatchFormat,
    templateId: state.selectedTemplateId,
  };
}

export type TemplateListState =
  | { status: "loading" }
  | { status: "populated"; templates: readonly MatchTemplate[]; isFromCache: boolean }
  | { status: "empty" }
  | { status: "fetch-failed-fallback-to-cache"; templates: readonly MatchTemplate[] }
  | { status: "fetch-failed-no-cache" };

/**
 * UX-04's own Error handling + Empty states, unified into one pure
 * function: "Template list fetch failure falls back to the
 * last-synced local cache silently if any exist, otherwise offers
 * 'Continue with custom setup' so the flow is never blocked."
 */
export function resolveTemplateListState(
  fetchSucceeded: boolean,
  fetchedTemplates: readonly MatchTemplate[] | null,
  cachedTemplates: readonly MatchTemplate[],
): TemplateListState {
  if (fetchSucceeded && fetchedTemplates !== null) {
    return fetchedTemplates.length === 0
      ? { status: "empty" }
      : { status: "populated", templates: fetchedTemplates, isFromCache: false };
  }
  if (cachedTemplates.length > 0) {
    return { status: "fetch-failed-fallback-to-cache", templates: cachedTemplates };
  }
  return { status: "fetch-failed-no-cache" };
}
